File: 03-spring-examples/apps/seda/02-webmvc-virtual-threads-rest-client/src/main/java/com/nalepa/demo/httpclient/HttpClientFactory.java
package com.nalepa.demo.httpclient;

import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

@Component
public class HttpClientFactory {

    private final RestClient.Builder restClientBuilder;

    public HttpClientFactory(RestClient.Builder restClientBuilder) {
        this.restClientBuilder = restClientBuilder;
    }

    public RestClient createRestClient() {
        PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
        connectionManager.setMaxTotal(200);
        connectionManager.setDefaultMaxPerRoute(200);

        var httpClient = HttpClients.custom().setConnectionManager(connectionManager).build();
        var requestFactory = new HttpComponentsClientHttpRequestFactory(httpClient);

        return restClientBuilder
                .requestFactory(requestFactory)
                .build();
    }
}

File: 03-spring-examples/helpers/request-sender/src/main/kotlin/com/nalepa/demo/httpclient/HttpClientFactory.kt
package com.nalepa.demo.httpclient

import org.apache.hc.client5.http.impl.classic.HttpClients
import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
import org.springframework.stereotype.Component
import org.springframework.web.client.RestClient

@Component
class HttpClientFactory(
    private val restClientBuilder: RestClient.Builder,
) {

    fun createRestClient(): RestClient {

        val connectionManager = PoolingHttpClientConnectionManager()
        connectionManager.maxTotal = 1000
        connectionManager.defaultMaxPerRoute = 1000
        val httpClient = HttpClients.custom().setConnectionManager(connectionManager).build()
        val requestFactory = HttpComponentsClientHttpRequestFactory(httpClient)

        return restClientBuilder
            .requestFactory(requestFactory)
            .build()
    }

}